test/e2e/install_test.go

```diff
@@ -93,7 +93,75 @@ var _ = Describe("Operator Install", func() {
 				g.Expect(bd.Status.Conditions[0].Reason).To(Equal("UnpackSuccessful"))
 				g.Expect(bd.Status.Conditions[1].Reason).To(Equal("InstallationSucceeded"))
 			}).WithTimeout(defaultTimeout).WithPolling(defaultPoll).Should(Succeed())
+		})
+		AfterEach(func() {
+			err := c.Delete(ctx, operatorCatalog)
+			Expect(err).ToNot(HaveOccurred())
+			err = c.Delete(ctx, operator)
+			Expect(err).ToNot(HaveOccurred())
+		})
+	})
+	When("resolving for an unavailable operator package", func() {
+		BeforeEach(func() {
+			ctx = context.Background()
+			pkgName = "argocd-operator"
+			operatorName = fmt.Sprintf("operator-%s", rand.String(8))
+			operator = &operatorv1alpha1.Operator{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: operatorName,
+				},
+				Spec: operatorv1alpha1.OperatorSpec{
+					PackageName: pkgName,
+				},
+			}
+			operatorCatalog = &catalogd.CatalogSource{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "test-catalog",
+				},
+				Spec: catalogd.CatalogSourceSpec{
+					// (TODO): Set up a local image registry, and build and store a test catalog in it
+					// to use in the test suite
+					Image: "quay.io/operatorhubio/catalog:latest",
+				},
+			}
+		})
+
+		It("resolves again when a new catalog is available", func() {
```

Comment:
I wonder if we can avoid all the long timeouts here by simplifying the test expectations. I think all we really want to assert is "whenever a catalog is created/updated/deleted, all of the existing operators are reconciled". Is there something that would show up in the `Operator` CR that we could use to verify that a reconcile happened after the catalog event?

That might be organized something like this:
```
Describe("reconcile Operators on catalog events")
   - BeforeEach() -> create 3 operators
   - AfterEach() -> delete 3 operators
   - When("catalog is created")
        - It("reconciles 3 operators")
   - When("catalog exists")
       - BeforeEach() -> create a catalog
       - AfterEach() -> delete the catalog
       - When("catalog is updated")
           - It("reconciles 3 operators")
       - When("catalog is deleted")
           - It("reconciles 3 operators")
```

Reply:
Or maybe even this is too much to do in the e2e? Is there a way to test the controller and the watches without a full-blown e2e?

If so, we could verify all these scenarios there, and then just focus on the happy path in the e2e?

I'm honestly not sure what's the best option here, but I think we should _really_ focus on keeping our overall e2e run as short (in terms of overall run time) as possible.

Reply:
I agree that it would be ideal if we could test this outside of the e2e tests, but as far as I am aware there isn't a good way to test watches being triggered without actually starting the controller and performing operations against a cluster that would trigger those watches. I think it *could* be done with envtest but would likely involve a lot of stubbing out resources and would essentially boil down to the question of "do we prefer complex unit tests or longer running e2e tests?"

Reply:
Using a minimal catalog cuts the wait down by a good chunk, so the e2e suite shouldn't be as long lived. I'm not sure if we'd want to expose resolution/reconcile details on the operator CR just for the sake of CI, in case it can be confusing to a user.

Reply:
If we're concerned about adding to the status API, how about examining the logs. We control the logs, and the tests can examine them for particular log entries. It wouldn't necessarily be exposing the logs as an API since it would be for e2e/unit tests, which can be modified when the logs are.

Reply:
The changes in this PR boil down to an updated controller configuration with a new watch that adds some extra stuff to the workqueue and calls a reconciler. So I'm suggesting it might be possible to isolate that specific behavior if we refactored a little bit (e.g. what if instead of `SetupWithManager` being a method on the reconciler, we made it a function that we could pass a reconciler to). In our `main.go`, we'd pass the real reconciler, but here, we'd pass a fake reconciler that does the "did I get triggered" assertions

Then we could put some envtest-based unit tests together alongside the reconciler and we would essentially split the testing into two chunks:
1. Does the controller configuration enqueue the expected reconcile requests?
2. Does the reconciler handle reconcile requests in the expected way?

We'd still want an e2e that runs through some scenarios of the integrated/real main.go, but we would avoid the combinatoric problem here, which would save _lots_ of time in e2e runs.

Reply:
> how about examining the logs

Yeah +1 if the unit test path isn't taken. We do this in the SDK helm-operator e2e tests.

Reply:
I'm opening another issue to track the refactoring https://github.com/operator-framework/operator-controller/issues/247. I don't think it makes sense to block this PR on the refactor, I'm in favor of having that in a separate PR.